File: crates/medousa-sdk/src/runtime.rs
#[cfg(feature = "async")]
use medousa_types::{
    AgentModeListResponse, AgentModeTransitionPolicy, ArtifactCommandRequest,
    ArtifactCommandResponse, ArtifactDeleteRequest, ArtifactDeleteResponse, ArtifactFetchRequest,
    ArtifactFetchResponse, ArtifactListUiRequest, ArtifactListUiResponse, ArtifactWriteRequest,
    ArtifactWriteResponse, RuntimeConfigCommandRequest, RuntimeConfigCommandResponse,
    StageRouteCommandRequest, StageRouteCommandResponse,
};

#[cfg(feature = "async")]
use crate::client::MedousaClient;
use crate::generated::ops;

#[cfg(feature = "async")]
pub struct RuntimeApi<'a> {
    pub(crate) client: &'a MedousaClient,
}

#[cfg(feature = "async")]
impl RuntimeApi<'_> {
    pub async fn agent_modes(&self) -> Result<AgentModeListResponse, crate::SdkError> {
        self.client.http().get(ops::AGENT_MODES_GET.path).await
    }

    pub async fn agent_mode_transition_policy(
        &self,
    ) -> Result<AgentModeTransitionPolicy, crate::SdkError> {
        self.client
            .http()
            .get(ops::AGENT_MODES_POLICY_GET.path)
            .await
    }

    pub async fn set_agent_mode_transition_policy(
        &self,
        policy: &AgentModeTransitionPolicy,
    ) -> Result<AgentModeTransitionPolicy, crate::SdkError> {
        self.client
            .http()
            .put(ops::AGENT_MODES_POLICY_PUT.path, policy)
            .await
    }

    pub async fn artifact_command(
        &self,
        request: &ArtifactCommandRequest,
    ) -> Result<ArtifactCommandResponse, crate::SdkError> {
        self.client
            .http()
            .post(ops::RUNTIME_ARTIFACT_COMMAND_POST.path, request)
            .await
    }

    pub async fn artifact_fetch(
        &self,
        request: &ArtifactFetchRequest,
    ) -> Result<ArtifactFetchResponse, crate::SdkError> {
        self.client
            .http()
            .post(ops::RUNTIME_ARTIFACT_FETCH_POST.path, request)
            .await
    }

    pub async fn artifact_list_ui(
        &self,
        request: &ArtifactListUiRequest,
    ) -> Result<ArtifactListUiResponse, crate::SdkError> {
        self.client
            .http()
            .post(ops::RUNTIME_ARTIFACT_LIST_UI_POST.path, request)
            .await
    }

    pub async fn artifact_write(
        &self,
        request: &ArtifactWriteRequest,
    ) -> Result<ArtifactWriteResponse, crate::SdkError> {
        self.client
            .http()
            .post(ops::RUNTIME_ARTIFACT_WRITE_POST.path, request)
            .await
    }

    pub async fn artifact_delete(
        &self,
        request: &ArtifactDeleteRequest,
    ) -> Result<ArtifactDeleteResponse, crate::SdkError> {
        self.client
            .http()
            .post(ops::RUNTIME_ARTIFACT_DELETE_POST.path, request)
            .await
    }

    pub async fn stage_route_command(
        &self,
        request: &StageRouteCommandRequest,
    ) -> Result<StageRouteCommandResponse, crate::SdkError> {
        self.client
            .http()
            .post(ops::RUNTIME_STAGE_ROUTE_COMMAND_POST.path, request)
            .await
    }

    pub async fn config_command(
        &self,
        request: &RuntimeConfigCommandRequest,
    ) -> Result<RuntimeConfigCommandResponse, crate::SdkError> {
        self.client
            .http()
            .post(ops::RUNTIME_CONFIG_COMMAND_POST.path, request)
            .await
    }
}

File: python/medousa-sdk/src/medousa/runtime.py
from __future__ import annotations

from medousa._decode import decode
from medousa._ops import op_path
from medousa.client import MedousaClient
from medousa.types import (
    ArtifactCommandRequest,
    ArtifactCommandResponse,
    ArtifactDeleteRequest,
    ArtifactDeleteResponse,
    ArtifactFetchRequest,
    ArtifactFetchResponse,
    ArtifactListUiRequest,
    ArtifactListUiResponse,
    ArtifactWriteRequest,
    ArtifactWriteResponse,
    RuntimeConfigCommandRequest,
    RuntimeConfigCommandResponse,
    StageRouteCommandRequest,
    StageRouteCommandResponse,
)


class RuntimeApi:
    def __init__(self, client: MedousaClient) -> None:
        self._client = client

    async def artifact_command(
        self,
        request: ArtifactCommandRequest,
    ) -> ArtifactCommandResponse:
        value = await self._client.transport.post_json(
            self._client.base_url,
            op_path("runtime.artifact.command.post"),
            request.model_dump(mode="json", exclude_none=True),
        )
        return decode(ArtifactCommandResponse, value)

    async def artifact_fetch(self, request: ArtifactFetchRequest) -> ArtifactFetchResponse:
        value = await self._client.transport.post_json(
            self._client.base_url,
            op_path("runtime.artifact.fetch.post"),
            request.model_dump(mode="json"),
        )
        return decode(ArtifactFetchResponse, value)

    async def artifact_write(self, request: ArtifactWriteRequest) -> ArtifactWriteResponse:
        value = await self._client.transport.post_json(
            self._client.base_url,
            op_path("runtime.artifact.write.post"),
            request.model_dump(mode="json", exclude_none=True),
        )
        return decode(ArtifactWriteResponse, value)

    async def artifact_delete(self, request: ArtifactDeleteRequest) -> ArtifactDeleteResponse:
        value = await self._client.transport.post_json(
            self._client.base_url,
            op_path("runtime.artifact.delete.post"),
            request.model_dump(mode="json", exclude_none=True),
        )
        return decode(ArtifactDeleteResponse, value)

    async def artifact_list_ui(self, request: ArtifactListUiRequest) -> ArtifactListUiResponse:
        value = await self._client.transport.post_json(
            self._client.base_url,
            op_path("runtime.artifact.list_ui.post"),
            request.model_dump(mode="json", exclude_none=True),
        )
        return decode(ArtifactListUiResponse, value)

    async def config_command(
        self,
        request: RuntimeConfigCommandRequest,
    ) -> RuntimeConfigCommandResponse:
        value = await self._client.transport.post_json(
            self._client.base_url,
            op_path("runtime.config.command.post"),
            request.model_dump(mode="json"),
        )
        return decode(RuntimeConfigCommandResponse, value)

    async def stage_route_command(
        self,
        request: StageRouteCommandRequest,
    ) -> StageRouteCommandResponse:
        value = await self._client.transport.post_json(
            self._client.base_url,
            op_path("runtime.stage_route.command.post"),
            request.model_dump(mode="json"),
        )
        return decode(StageRouteCommandResponse, value)
